File: ydb/benches/sdk_compare/payload.rs
use anyhow::{Result, ensure};

pub(crate) const HEADER_SIZE_BYTES: usize = 8;

const FILL_BYTE: u8 = 0xA5;

pub(crate) fn allocate(message_size_bytes: usize) -> Result<Vec<u8>> {
    ensure!(
        message_size_bytes >= HEADER_SIZE_BYTES,
        "message size must be at least {HEADER_SIZE_BYTES} bytes"
    );

    let mut payload = vec![FILL_BYTE; message_size_bytes];
    payload[..HEADER_SIZE_BYTES].fill(0);
    Ok(payload)
}

pub(crate) fn write_timestamp(payload: &mut [u8], sent_at_ns: u64) -> Result<()> {
    ensure!(
        payload.len() >= HEADER_SIZE_BYTES,
        "payload is shorter than the {HEADER_SIZE_BYTES}-byte header"
    );
    payload[..HEADER_SIZE_BYTES].copy_from_slice(&sent_at_ns.to_le_bytes());
    Ok(())
}

pub(crate) fn read_timestamp(payload: &[u8]) -> Result<u64> {
    ensure!(
        payload.len() >= HEADER_SIZE_BYTES,
        "payload is shorter than the {HEADER_SIZE_BYTES}-byte header"
    );

    let mut sent_at_ns = [0_u8; 8];
    sent_at_ns.copy_from_slice(&payload[..HEADER_SIZE_BYTES]);
    Ok(u64::from_le_bytes(sent_at_ns))
}

File: benchmarks/sdk-compare/cpp/src/core/payload.cpp
#include "payload.h"

#include <stdexcept>

#include "scenario.h"

namespace sdk_compare {
namespace {

inline constexpr unsigned char fill_byte = 0xA5;

void require_header(std::size_t size, std::string_view subject) {
    if (size < topic_payload_header_size_bytes) {
        throw std::runtime_error(
            std::string(subject) + " is shorter than the " +
            std::to_string(topic_payload_header_size_bytes) + "-byte header");
    }
}

}  // namespace

std::string allocate_payload(std::size_t message_size_bytes) {
    require_header(message_size_bytes, "message size");
    std::string payload(message_size_bytes, static_cast<char>(fill_byte));
    payload.replace(0,
                    topic_payload_header_size_bytes,
                    topic_payload_header_size_bytes,
                    '\0');
    return payload;
}

void write_timestamp(std::string& payload, std::uint64_t sent_at_ns) {
    require_header(payload.size(), "payload");
    for (std::size_t index = 0; index < topic_payload_header_size_bytes;
         ++index) {
        payload[index] =
            static_cast<char>((sent_at_ns >> (index * 8U)) & 0xFFU);
    }
}

std::uint64_t read_timestamp(std::string_view payload) {
    require_header(payload.size(), "payload");
    std::uint64_t sent_at_ns = 0;
    for (std::size_t index = 0; index < topic_payload_header_size_bytes;
         ++index) {
        const auto byte = static_cast<unsigned char>(payload[index]);
        sent_at_ns |= static_cast<std::uint64_t>(byte) << (index * 8U);
    }
    return sent_at_ns;
}

}  // namespace sdk_compare
